beetsplug/replaygain.py

```diff
@@ -957,24 +976,108 @@ def join(self, timeout=None):
+        self.album_gain = None
+        self.track_gains = None
+
+    def _store_track_gain(self, item, track_gain):
+        item.rg_track_gain = track_gain.gain
+        item.rg_track_peak = track_gain.peak
+        item.store()
+        self._log.debug(u'applied track gain {0} LU, peak {1} of FS',
+                        item.rg_track_gain, item.rg_track_peak)
+
+    def _store_album_gain(self, item):
+        """
+
+        The caller needs to ensure that `self.album_gain is not None`.
+        """
+        item.rg_album_gain = self.album_gain.gain
+        item.rg_album_peak = self.album_gain.peak
+        item.store()
+        self._log.debug(u'applied album gain {0} LU, peak {1} of FS',
+                        item.rg_album_gain, item.rg_album_peak)
+
+    def store_track(self, write):
+        item = self.items[0]
+        if self.track_gains is None or len(self.track_gains) != 1:
+            # In some cases, backends fail to produce a valid
+            # `track_gains` without throwing FatalReplayGainError
+            #  => raise non-fatal exception & continue
+            raise ReplayGainError(
+                u"ReplayGain backend `{}` failed for track {}"
+                .format(self.backend_name, item)
+            )
+
+        self._store_track_gain(item, self.track_gains[0])
+        if write:
+            item.try_write()
+        self._log.debug(u'done analyzing {0}', item)
+
+    def store_album(self, write):
+        if (self.album_gain is None or self.track_gains is None
+                or len(self.track_gains) != len(self.items)):
+            # In some cases, backends fail to produce a valid
+            # `album_gain` without throwing FatalReplayGainError
+            #  => raise non-fatal exception & continue
+            raise ReplayGainError(
+                u"ReplayGain backend `{}` failed "
+                u"for some tracks in album {}"
+                .format(self.backend_name, self.album)
+            )
+        for item, track_gain in zip(self.items, self.track_gains):
+            self._store_track_gain(item, track_gain)
+            self._store_album_gain(item)
+            if write:
+                item.try_write()
+            self._log.debug(u'done analyzing {0}', item)
+
+    def store(self, write):
+        if self.album is not None:
+            self.store_album(write)
+        else:
+            self.store_track(write)
+
+
+class R128Task(RgTask):
+    def _store_track_gain(self, item, track_gain):
+        item.r128_track_gain = track_gain.gain
+        item.store()
+        self._log.debug(u'applied r128 track gain {0} LU',
+                        item.r128_track_gain)
+
+    def _store_album_gain(self, item):
+        """
+
+        The caller needs to ensure that `self.album_gain is not None`.
+        """
+        item.r128_album_gain = self.album_gain.gain
+        item.store()
+        self._log.debug(u'applied r128 album gain {0} LU',
+                        item.r128_album_gain)
+
```

Comment:
I'd put `Task` and `RgTask` in front of `Backend` & subclasses since they're used in those.
And maybe add some docstrings to the 'public' methods?

Reply:
Good point; done!